src/server.rs

```diff
@@ -0,0 +1,450 @@
+        RustboxConfig {
+            services: src_list,
+            fxa_host: String::from(
+                config
+                    .get_str("fxa_host")
+                    .unwrap_or("oauth.stage.mozaws.net"),
+            ),
+            dryrun: config.get_bool("dryrun").unwrap_or(false),
+            default_ttl: config.get_float("default_ttl").unwrap_or(3600.0) as i64,
+            test_data: config
+                .get_table("test_data")
+                .unwrap_or(&Table::new())
+                .clone(),
+        }
+    }
+}
+
+impl<'a, 'r> FromRequest<'a, 'r> for RustboxConfig {
+    type Error = ();
+
+    fn from_request(req: &'a Request<'r>) -> request::Outcome<Self, ()> {
+        Outcome::Success(req.guard::<State<RustboxConfig>>().unwrap().inner().clone())
+    }
+}
+
+// Encapsulate the server.
+pub struct Server {}
+
+impl Server {
+    pub fn start(rocket: rocket::Rocket) -> HandlerResult<rocket::Rocket> {
+        Ok(rocket
+            .attach(AdHoc::on_attach(|rocket| {
+                // Copy the config into a state manager.
+                let pool = pool_from_config(rocket.config()).expect("Could not get pool");
+                let rbconfig = RustboxConfig::new(rocket.config());
+                Ok(rocket.manage(rbconfig).manage(pool))
+            }))
+            .mount(
+                "/v1/store",
+                routes![read, read_opt, write, delete, delete_user],
+            )
+            .mount("/v1/", routes![status]))
+    }
+}
+
+// Method handlers:::
+// Apparently you can't set these on impl methods, must be at top level.
+//  query string parameters for limit and index
+#[get("/<service>/<user_id>/<device_id>?<options>")]
+fn read_opt(
+    conn: Conn,
+    token: HandlerResult<FxAAuthenticator>,
+    service: String,
+    user_id: String,
+    device_id: String,
+    options: Options,
+) -> HandlerResult<Json> {
+    // 👩🏫 note that the "token" var is a HandlerResult wrapped Validate struct.
+    // Validate::from_request extracts the token from the Authorization header, validates it
+    // against FxA and the method, and either returns OK or an error. We need to reraise it to the
+    // handler.
+
+    if token.is_err() {
+        return Err(token.err().unwrap());
+    }
+    let max_index = DatabaseManager::max_index(&conn, &user_id, &device_id, &service);
+    let index = options.index.unwrap_or(0 as i64);
+    let limit = options.limit.unwrap_or(0 as i64);
+    let messages =
+        DatabaseManager::read_records(&conn, &user_id, &device_id, &service, index, limit).unwrap();
+    // returns json {"status":200, "index": max_index, "messages":[{"index": #, "data": String}, ...]}
+    let mut is_last = true;
+    if limit > 0 && messages.len() == limit as usize {
+        let last = messages.last().unwrap();
+        is_last = last.idx == max_index;
+    }
+    Ok(Json(json!({
+        "last": is_last,
+        "index": max_index.clone(),
```

Comment:
isn't "index" returned here the last index in the *current* batch we are returning to the client? So it knows where it left off. max_index here is the last index of *all messages*, which isn't right.